File: 프로그래머스/JAVA/소수찾기.java
import java.util.*;
class Solution {
    
    static int size;
    static StringBuilder sb;
    static HashSet<Integer> hashset;
    
    public int solution(String numbers) {
        int answer = 0;
        size = numbers.length();
        hashset = new HashSet<>();
        
        for(int i = 1; i<=size; i++){
            sb = new StringBuilder();
            sb.append(numbers);
            perm(i, 0);
        }
        answer = hashset.size();
        return answer;
    }
    
    public static boolean sosu(int a){
        if(a == 1 || a == 0){
            return false;
        }
        if(a == 2){
            return true;
        }
        for(int i = 2; i*i<=a; i++){
            if(a%i == 0)
                return false;
        }
        return true;
    }
    public static void swap(int x, int y){
        char tmp1 = sb.charAt(x);
        char tmp2 = sb.charAt(y);
        sb.setCharAt(x,tmp2);
        sb.setCharAt(y,tmp1);
    }
    public static void perm(int step, int start){
        if(start == step){
            int a = Integer.parseInt(sb.toString());
            int sub = (int)Math.pow(10,(size-step));
            
            a = a/sub;
            // System.out.println(a);
            if(sosu(a)){
                hashset.add(a);
            }
            return;
        }
        
        for(int i = start; i<size; i++){
            swap(i,start);
            perm(step, start+1);
            swap(i,start);
        }
    }
}

File: 프로그래머스/C++/소수찾기.cpp
#include <string>
#include <vector>
#include <set>
#include <iostream>
using namespace std;

bool visited[7];
int arr[7];
set<int> s;

bool sosu(int num){
    if(num == 0 || num == 1){
        return false;
    }
    if(num == 2){
        return true;
    }
    
    for(int i = 2; i*i<=num; i++){
        if(num % i == 0)
            return false;
    }
    
    return true;
}


void permu(int finish, int step, string numbers){
    if(step == finish){
        
        int num = 0;
        
        for(int i = 0; i<step; i++){
            num = num*10 + arr[i];
        }
        if(sosu(num))
            s.insert(num);
        
        return;
    }
    
    for(int i = 0; i<numbers.size(); i++){
        if(visited[i])
            continue;
        visited[i] = true;
        arr[step] = numbers[i]-'0';
        permu(finish, step+1, numbers);
        visited[i] = false;
        arr[step] = 0;
    }
}

int solution(string numbers) {
    int answer = 0;
    
    for(int i = 1; i<=numbers.size(); i++){
        fill_n(visited, 7, false);
        fill_n(arr, 7, 0);
        permu(i, 0, numbers);
    }
    
    answer = s.size();
    return answer;
}
